hibernate-core/src/main/java/org/hibernate/jpa/event/internal/CallbackRegistryImpl.java

```diff
@@ -6,51 +6,57 @@
  */
 package org.hibernate.jpa.event.internal;
 
-import java.util.HashMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
 import jakarta.persistence.PersistenceException;
 
 import org.hibernate.internal.util.collections.ArrayHelper;
+import org.hibernate.internal.util.collections.MapBackedClassValue;
 import org.hibernate.jpa.event.spi.Callback;
+import org.hibernate.jpa.event.spi.CallbackRegistry;
 import org.hibernate.jpa.event.spi.CallbackType;
 
 /**
  * Keep track of all lifecycle callbacks and listeners for a given persistence unit
  *
  * @author <a href="mailto:kabir.khan@jboss.org">Kabir Khan</a>
  * @author Steve Ebersole
+ * @author Sanne Grinovero
  */
-final class CallbackRegistryImpl implements CallbackRegistryImplementor {
-	private final HashMap<Class<?>, Callback[]> preCreates = new HashMap<>();
-	private final HashMap<Class<?>, Callback[]> postCreates = new HashMap<>();
-	private final HashMap<Class<?>, Callback[]> preRemoves = new HashMap<>();
-	private final HashMap<Class<?>, Callback[]> postRemoves = new HashMap<>();
-	private final HashMap<Class<?>, Callback[]> preUpdates = new HashMap<>();
-	private final HashMap<Class<?>, Callback[]> postUpdates = new HashMap<>();
-	private final HashMap<Class<?>, Callback[]> postLoads = new HashMap<>();
+final class CallbackRegistryImpl implements CallbackRegistry {
+
+	private final MapBackedClassValue<Callback[]> preCreates;
+	private final MapBackedClassValue<Callback[]> postCreates;
+	private final MapBackedClassValue<Callback[]> preRemoves;
+	private final MapBackedClassValue<Callback[]> postRemoves;
+	private final MapBackedClassValue<Callback[]> preUpdates;
+	private final MapBackedClassValue<Callback[]> postUpdates;
+	private final MapBackedClassValue<Callback[]> postLoads;
+
+	public CallbackRegistryImpl(
+			ConcurrentMap<Class<?>, Callback[]> preCreates,
+			ConcurrentMap<Class<?>, Callback[]> postCreates,
+			ConcurrentMap<Class<?>, Callback[]> preRemoves,
+			ConcurrentMap<Class<?>, Callback[]> postRemoves,
+			ConcurrentMap<Class<?>, Callback[]> preUpdates,
+			ConcurrentMap<Class<?>, Callback[]> postUpdates,
+			ConcurrentMap<Class<?>, Callback[]> postLoads) {
```

Comment:
Maybe I'm missing something, but it seems the `Builder` is not used concurrently and the only other method that mutates maps is `CallbackRegistryImpl#release` which is called by `EventEngine#stop`, though that method has no caller. So we could use a plain `HashMap` as well, right? Apart from that, is there any reason for clearing the maps? The `EventEngine` and hence the `CallbackRegistryImpl` will become unreachable as soon as the `SessionFactoryImpl` becomes unreachable, so clearing the maps seems a bit moot to me, especially since it seems to require the use of a concurrent map.

Reply:
>  So we could use a plain HashMap as well, right?

No, while this Map isn't going to mutated it's still going to be *read* concurrently and that requires a ConcurrentMap. BTW a `ConcurrentHashMap` outperforms a plain `HashMap` in such use case as well.

> Apart from that, is there any reason for clearing the maps? 

I wondered about that myself and wasn't entirely sure. It seems previously it was very careful with these references so I tried to maintain that. I suppose it's possible that we get in a reference circularity, which gets quite bad when there's Class types and classloaders involved.